library/webpack-plugins/react-universal-plugin.js

```diff
@@ -1,96 +1,204 @@
-            if (!duplicate) additionalEntries.push(dep)
-
-          } else if (query === '?universal-client') {
-            loaders.push({ loader: require.resolve('../webpack-loaders/react-universal-client-loader') })
+            if (!clientEntries[name]) clientEntries[name] = './' + name + '?universal-client'
           }
+
           done(null, data)
         })
+      })
 
-        // mark client entry modules and provide `universal` variables
-        compilation.plugin('normal-module-loader', (context, module) => {
-          module.universalClient = module.userRequest.endsWith('?universal-client')
-        })
+      // tell the web compiler to compile, emit the assets and notify the appropriate plugins
+      compiler.plugin('make-additional-entries', (compilation, createEntries, done) => {
+
+        const startTime = Date.now()
+        webCompiler.compile((err, webCompilation) => {
+          if (err) return finish(err)
+
+          compilation.children.push(webCompilation)
 
-        // remove redundant assets introduced by client chunk
-        compilation.plugin('after-optimize-chunk-assets', chunks => {
-          chunks.forEach(chunk => {
-            const { entryModule, files, name } = chunk
-            if (entryModule.universalClient) {
-              Object.keys(compilation.assets).forEach(assetName => {
-                if (assetName != name && assetName.startsWith(name)) {
-                  delete compilation.assets[assetName]
-                }
-              })
-            }
+          webCompiler.emitAssets(webCompilation, err => {
+            if (err) return finish(err)
+
+            finish(null, webCompilation)
           })
         })
-      })
 
-      // `make` all `initialEntries` and add `additionalEntries` when they have been recorded
-      compiler.plugin('make', (compilation, done) => {
-        Promise.all(initialEntries.map(addEntry))
-          .then(_ => Promise.all(additionalEntries.map(addEntry)))
-          .then(_ => { done() })
-          .catch(e => { done(e) })
+        function finish(err, compilation) {
+          if (err) {
+            webCompiler.applyPlugins('failed', err)
+            return done(err)
+          }
+
+          const stats = new Stats(compilation)
+          stats.startTime = startTime
+          stats.endTime = Date.now()
 
-        function addEntry(entry) {
-          return new Promise((resolve, reject) => {
-            compilation.addEntry(compiler.context, entry, entry.loc, err => err ? reject(err) : resolve())
-          })
+          webCompiler.applyPlugins('done', stats)
+
+          done()
         }
       })
     }
   }
 }
+
+function createWebCompiler(compiler, getEntries) {
+
+  // Massage the options to become a web configuration
+  const options = Object.assign({}, compiler.options)
+  options.name = 'react-universal-plugin-client-compiler'
+  options.target = 'web'
+  options.entry = undefined
+  options.externals = undefined
```

Comment:
consider moving these to a `targetBasedOptionsPlugin`

Reply:
When I wrote the previous comment I couldn't remember why I removed a plugin that did just that. The reason for not having that as a plugin is because the options are passed to a compiler before any plugin is run. So keep it 'clean' and use options as options. We could however think about removing them from the base (node target) options.

Hmmm